File: generated_code/cpp/solution_451_2.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    string frequencySort(string s) {
        // Step 1: Count the frequency of each character.
        unordered_map<char, int> counts;
        for (char c : s) {
            counts[c]++;
        }
        
        // Step 2: Use bucket sort for an optimal O(N) time complexity.
        // The maximum frequency a character can have is s.length().
        int n = s.length();
        vector<vector<char>> buckets(n + 1);
        for (auto& [ch, freq] : counts) {
            buckets[freq].push_back(ch);
        }
        
        // Step 3: Build the result string by iterating from the highest frequency bucket down to 1.
        string result;
        result.reserve(n); // Reserve memory to avoid reallocations.
        for (int freq = n; freq >= 1; --freq) {
            for (char ch : buckets[freq]) {
                result.append(freq, ch);
            }
        }
        
        return result;
    }
};

File: generated_code/python/solution_451_2.py
from collections import Counter
from typing import List

class Solution:
    def frequencySort(self, s: str) -> str:
        # Step 1: Count the frequency of each character
        # Time Complexity: O(N) where N is the length of the string
        # Space Complexity: O(K) where K is the number of unique characters (K <= 62 for letters/digits)
        counts = Counter(s)
        
        # Step 2: Group characters by their frequency using bucket sort
        # Since the maximum frequency cannot exceed the length of the string,
        # we can use an array of lists where index represents the frequency.
        max_freq = len(s)
        buckets: List[List[str]] = [[] for _ in range(max_freq + 1)]
        
        for char, freq in counts.items():
            buckets[freq].append(char)
            
        # Step 3: Iterate through the buckets in reverse order (highest frequency first)
        # and reconstruct the result string.
        # Overall Time Complexity: O(N)
        # Overall Space Complexity: O(N) to store the output elements
        result = []
        for freq in range(max_freq, 0, -1):
            for char in buckets[freq]:
                result.append(char * freq)
                
        return "".join(result)